src/state/farms/selectors.ts

```diff
@@ -0,0 +1,106 @@
+import BigNumber from 'bignumber.js'
+import { BIG_ZERO } from 'utils/bigNumber'
+import { getBalanceAmount } from 'utils/formatBalance'
+import { createSelector } from '@reduxjs/toolkit'
+import { State, SerializedFarm, DeserializedFarm, DeserializedFarmUserData } from '../types'
+import { deserializeToken } from '../user/hooks/helpers'
+
+const deserializeFarmUserData = (farm: SerializedFarm): DeserializedFarmUserData => {
+  return {
+    allowance: farm.userData ? new BigNumber(farm.userData.allowance) : BIG_ZERO,
+    tokenBalance: farm.userData ? new BigNumber(farm.userData.tokenBalance) : BIG_ZERO,
+    stakedBalance: farm.userData ? new BigNumber(farm.userData.stakedBalance) : BIG_ZERO,
+    earnings: farm.userData ? new BigNumber(farm.userData.earnings) : BIG_ZERO,
+  }
+}
+
+const deserializeFarm = (farm: SerializedFarm): DeserializedFarm => {
+  const { lpAddresses, lpSymbol, pid, dual, multiplier, isCommunity, quoteTokenPriceBusd, tokenPriceBusd } = farm
+
+  return {
+    lpAddresses,
+    lpSymbol,
+    pid,
+    dual,
+    multiplier,
+    isCommunity,
+    quoteTokenPriceBusd,
+    tokenPriceBusd,
+    token: deserializeToken(farm.token),
+    quoteToken: deserializeToken(farm.quoteToken),
+    userData: deserializeFarmUserData(farm),
+    tokenAmountTotal: farm.tokenAmountTotal ? new BigNumber(farm.tokenAmountTotal) : BIG_ZERO,
+    lpTotalInQuoteToken: farm.lpTotalInQuoteToken ? new BigNumber(farm.lpTotalInQuoteToken) : BIG_ZERO,
+    lpTotalSupply: farm.lpTotalSupply ? new BigNumber(farm.lpTotalSupply) : BIG_ZERO,
+    tokenPriceVsQuote: farm.tokenPriceVsQuote ? new BigNumber(farm.tokenPriceVsQuote) : BIG_ZERO,
+    poolWeight: farm.poolWeight ? new BigNumber(farm.poolWeight) : BIG_ZERO,
+  }
+}
+
+const selectCakeFarm = (state: State) => state.farms.data.find((f) => f.pid === 251)
+const selectFarmFromPid = (pid: number) => (state: State) => state.farms.data.find((f) => f.pid === pid)
```

Comment:
I think we could consolidate `selectFarmFromPid` and `selectFarmFromLpSymbol` into one function.
```
const selectFarmByKey = (key: string, value: string|number) => data.find(f => f[key] === value)
```

Reply:
Also, should we add `deserializeFarm` in this function as well ?

Reply:
> `const selectFarmByKey = (key: string, value: string|number) => data.find(f => f[key] === value)`

Done

Reply:
> Also, should we add `deserializeFarm` in this function as well ?

We cannot put deserializeFarm inside the selector, it will break memoization because it creates a new object